### DISCUSSION TITLE
Scalability - Universal access to Istio and Kubernetes resources

### DISCUSSION THREAD

**Discussion by xeviknal** [kiali/kiali#4515] [Roadmap]
In my opinion we should have an snapshot in memory of all the resources we need to build our features. This snapshot should be properly cached and properly expired (at resource granularity - e.g. Service has been update -> update the service in the snapshot).

The context of this is the following one. Kiali validations need a high volume of resources (amount of instances and kinds). Those resources are consumed from different namespaces, in cross namespaces fashion, and need to be available quickly to be perforant in different demanding situations.

There is one feature coming: [Adding object references to Istio objects in the side panel ](https://github.com/kiali/kiali/issues/4241). This feature will need, again, access to the whole service mesh resources in order to link fields from an Istio object to any other linkable resource.

This is one example:

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: service-b-destination-rule
spec:
  host: reviews
  subsets:
  - name: v1
    labels:
      - app: reviews
        version: v1
  - name: v2
    labels:
      - app: reviews
        version: v2
```

The references should be something like:
```yaml
references: [
{ type: service, name: reviews, namespace: bookinfo, meshId: blahMesh }
{ type: workload, name: reviews-v1, namespace: bookinfo, meshId: blahMesh }
{ type: workload, name: reviews-v2, namespace: bookinfo, meshId: blahMesh }
]
```

Potentially, there could be linked things like WorkloadEntries, ServiceEntries, Sidecars, and Apps, Workloads, Services, and many more.

As the validations are now, we are always fetching all the istio resources needed for all the validations even those that are not necessary for one specific validation. All the fetches are hitting the cache, however, there are way too many fetches to do to match what is needed to compute a validation or validation of all istio objects of one namespace.

Having an snapshot of the needed resources would remove the pain of call those fetches in every single request. Of course, if validations are cached, this should remove all the fetches pain as well. However, re populating the cache would be painful as well. Plus, this snapshot would be very useful to compute the references of an istio object (#4241).

This bit here is the one performing quite a lot of fetches:
https://github.com/xeviknal/kiali/blob/06ab1bd6ded54622812ce74a8af26d7ca0823db2/business/istio_validations.go#L69-L77

This bit here is how the istioctl analyzer maintains the cache:

https://github.com/istio/istio/blob/d0748d7c46c02e1fddf77d7a38f90cb3620e4358/pkg/config/analysis/local/context.go#L47-L54

That implements this interface:
https://github.com/istio/istio/blob/d0748d7c46c02e1fddf77d7a38f90cb3620e4358/pkg/config/analysis/context.go#L27-L42

**Comment by lucasponce** [comment]
I think with the work being proposed in https://github.com/kiali/kiali/pull/4479 Kiali would move into that direction and goal of https://github.com/kiali/kiali/issues/4426 is to use this cached information.

To introduce a new index of references is a good idea, it will help to move some logic from O(N) -> O(1) and that would have a positive aspect in large deployments.
